lib/workers.js

```diff
@@ -0,0 +1,222 @@
+const { EventEmitter } = require('events');
+const { Worker } = require('worker_threads');
+const Codecept = require('./codecept');
+const Container = require('./container');
+const { getTestRoot } = require('./command/utils');
+const mainConfig = require('./config');
+const output = require('./output');
+const path = require('path');
+const event = require('./event');
+const { Suite, Test } = require('mocha');
+
+const pathToWorker = path.join(__dirname, 'command', 'workers', 'runTests.js');
+
+const initializeCodecept = (configPath) => {
+  const codecept = new Codecept(mainConfig.load(configPath), {});
+  codecept.init(getTestRoot(configPath));
+  codecept.loadTests();
+
+  return codecept;
+};
+
+const populateGroups = (numberOfWorkers) => {
+  const groups = [];
+  for (let i = 0; i < numberOfWorkers; i++) {
+    groups[i] = [];
+  }
+
+  return groups;
+};
+
+const createWorker = (workerObject) => {
+  return new Worker(pathToWorker, {
+    workerData: {
+      options: simplifyObject(workerObject.options),
+      tests: workerObject.tests,
+      testRoot: workerObject.testRoot,
+      workerIndex: workerObject.workerIndex + 1,
+    },
+  });
+};
+
+const simplifyObject = (object) => {
+  return Object.keys(object)
+    .filter(k => k.indexOf('_') !== 0)
+    .filter(k => typeof object[k] !== 'function')
+    .filter(k => typeof object[k] !== 'object')
+    .reduce((obj, key) => {
+      obj[key] = object[key];
+      return obj;
+    }, {});
+};
+
+const repackTest = (test) => {
+  test = Object.assign(new Test(test.title || '', () => { }), test);
+  test.parent = Object.assign(new Suite(test.parent.title), test.parent);
+  return test;
+};
+
+const createWorkerObjects = (testGroups, config) => {
+  return testGroups.map((tests, index) => new WorkerObject({ config: JSON.stringify(config) }, tests, getTestRoot(), index));
+};
+
+const indexOfSmallestElement = (groups) => {
+  let i = 0;
+  for (let j = 1; j < groups.length; j++) {
+    if (groups[j - 1].length > groups[j].length) {
+      i = j;
+    }
+  }
+  return i;
+};
+
+class WorkerObject {
+  constructor(options, tests, testRoot, workerIndex) {
+    this.options = options || {};
+    this.tests = tests;
+    this.testRoot = testRoot || getTestRoot();
+    this.workerIndex = workerIndex;
+  }
+
+  addConfig(config) {
+    this.options.config = JSON.stringify(config);
+  }
+
+  addTests(tests) {
+    this.tests = tests;
+  }
+}
+
+class Workers extends EventEmitter {
+  /**
+   * @param {Number} numberOfWorkers
+   * @param {Boolean} runAsSuite
+   */
+  constructor(numberOfWorkers, runAsSuite, configPath) {
+    super();
+    this.codecept = initializeCodecept(configPath);
+    this.finishedTests = [];
+    this.errors = [];
+    this.numberOfWorkers = 0;
+    this.closedWorkers = 0;
+    this.workers = [];
+    this.stats = {
+      passes: 0,
+      failures: 0,
+      tests: 0,
+    };
+
+    if (typeof numberOfWorkers === 'number' && numberOfWorkers > 0) {
```

Comment:
this part should be moved into `spawn()`, also `spawn` as options should take `runAsSuite` or `runAsGroup` as argument with a default value for 'tests'. Alternatively, `spawn` can take a function which could split tests by its own. So if a user needs to split tests by their speed, they can use that function. 

So signature should be: 

* `spawn('test')` same as `spawn()`
* `spawn('suite')`
* `spawn((number) => [ /** groups **/ ])`

Reply:
@DavertMik I have a  small doubt here. I thought spawn is for creating a custom worker. For simple worker operation, we can use something like this:
```js
const workers = new Workers(2, false);
workers.run();
``` 
And we can use spawn for creating workers with custom configs. I may be wrong here. Your view on this?.